### PR TITLE
Speed up Steerable Pyramid, Portilla-Simoncelli

### LINKED ISSUE #222: Make Portilla-Simoncelli code more efficient
Because of the multiscale representation used in the Portilla-Simoncelli texture model, it's hard to write the code in a GPU-friendly way. I attempted to do so with in [this commit](https://github.com/LabForComputationalVision/plenoptic/blob/19cf5217f5cd2ba8842b6c3573fae6ba1345b650/src/plenoptic/simulate/models/portilla_simoncelli.py), by using the non-downsampled pyramid, which allowed me to store the coefficients as a single tensor of shape `(batch, channel, n_scales, n_orientations, height, width)`, which made a variety of things easier. It is probably worth comparing against the version [here](https://github.com/LabForComputationalVision/plenoptic/blob/9c7f2dba10caa4f814a95f0355035026919a2a82/src/plenoptic/simulate/models/portilla_simoncelli.py), which uses the downsampled version, packs the coefficients into lists of length `n_scales`, each entry of which is a tensor of shape `(batch, channel, n_orientations, height/2^s, width/2^s)`, where `s` is the scale / index in the list, and uses list comprehension for much of the computations.

However, the version that uses the non-downsampled pyramid is *significantly* slower on the CPU (slightly faster on the GPU), much more memory-intensive, and it's *really* hard to compare against the earlier code -- in general, it's just not the same, because e.g., computing the autocorrelation of the coefficients at the coarsest scale gives a different answer depending on when that scale is size `(64, 64)` or `(256, 256)`. You can downsample the coefficient image before computing the autocorrelation or downsampling it afterwards, but there's no way to do *that* efficiently on the GPU ([vmap doesn't let you convert something being vmapped over to an int](https://github.com/google/jax/discussions/8973?sort=top) or [like dynamically-sized inputs](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#dynamic-shapes), one of which would be necessary to do the downsampling and center cropping in an efficient manner) and it's still generally not the same value (something like `allclose(rtol=1e-1, atol=1e-1)`). And you must do the downsampling, or the definition of "autocorrelations up to 4 shifts in all directions" is completely different at the coarser scales.

With the downsampled + list comprehension version of the code, I tried considering the lists as `pytree` and using `tree_map` (either from `torch.utils._pytree`, `jax`, or [optree](https://github.com/metaopt/optree)) and got no speedup (either on GPU or CPU, probably because a list of 4d/5d tensors is not that difficult of a pytree to parse), and vmap doesn't like [ragged data](https://github.com/google/jax/discussions/5184) (a list of variably-sized arrays) or [lists of structs](https://github.com/google/jax/discussions/14021), so I couldn't figure out a way to make anything vmap-able.

So, unsure how to improve this.

### LINKED ISSUE #233: Refactor steerable pyramid
Refactor `SteerablePyramidFreq`, similar to the `PortillaSimoncelli` refactor in #225. Like the `PortillaSimoncelli` code before that PR, the steerable pyramid code is more or less a straight port from the matlab (by way of [pyrtools](https://github.com/LabForComputationalVision/pyrtools/)). It would be nice to refactor, clean it up, and speed it up. 

Copying Edardo's comments from that. All these comments are in `__init__`:

> Line 50, about the docstring for `twidth`: I don't love this compact name, something like transition_width would be more explicit

> Line 61, about the explanation of tight frame: In the attributes section, it looks like you have some attribute missing: order, downsample, tight_frame, alpha, lutsize... some are listed as Parameters and some are not. if this is by choice because you do not want to replicate the docstrings of the Parameters of the__init__, you should remove is_complex and image_shape) because they are redundant.

> Line 118, about definition of `self.Xcosn`: what are this for? Xcosn and alpha? it would be nice to have a description of those to make the code more accessible to a new contributor/developer

> Line 146, about `(xramp, yramp)` definition: there are a lot attributes, probably tedious to fill in all the info, but it would be cool to have a short description of what are those and why we are computing them.

> Line 156, about the call to `raised_cosine`: np.ndarray doesn't match signature of raised_cosine, which wants Tuple[float, float]
> Suggested change
>   `-` self.Xrcos, Yrcos = raised_cosine(twidth, (-twidth / 2.0), np.array([0, 1]))
>   `+` self.Xrcos, Yrcos = raised_cosine(twidth, (-twidth / 2.0), (0.,1.))
> I just checked the raised cosine function, here it looks like you are squaring the cosine, which is different from what we did for the the GLM basis; In the GLM basis we did use the same formula that was in the original pillow paper and codebase. This is from https://github.com/pillowlab/raisedCosineBasis/blob/master/makeRaisedCosBasis.m
> `raisedCosFun = @(x,ctr,dCtr)((cos(max(-pi,min(pi,(x-ctr)*pi/dCtr/2)))+1)/2);'`
> which is (cos(*) +1)/2, and * is truncated to only one bump.
> I am not saying you should change what is done here, but I see it can create confusion

This is related to https://github.com/flatironinstitute/log-stretched-basis/issues/2 I think, showing that the various parametrizations are equivalent or at least explaining the relationship between them.

> Line 183, in `__init__` about `self.scales`: you'll define a self.scales in the portilla_simoncelli.py too, but the two definition seems slightly different, since self.pyr is an attribute of the PortillaSimoncelli class, is it worth having the scales defined only here, in the steerable pyramid class?

As I commented in #225, they're different because `self.scales` is model-specific, and the Portilla-Simoncelli texture model additionally considers marginal pixel statistics (first several moments). However, `self.scales` is only used for coarse-to-fine optimization, and the pyramid generally should not be used as a model by itself (which is why it's in the `canonical_computations/` directory instead of `models/`). So should the pyramid's scales attribute be removed?

> Line 196, in `__init__` about `const`: can you compute this one outside the loop?

> Line 212, in `__init__` about `Ycosn_forward`+`Ycosn_recon`: you are using const, self.Xcosn, slef.alpha, self.order, which looks all independent from "i" so you may move this outside the loop too. 

> general feeling, there are a lot of variables with similar names, Xrcos, Xcosn, is there a way to find names that are more intuitive?

### PR DESCRIPTION
**Pull request by billbrod**
**Describe the change in this PR at a high-level**

As part of the investigation into PortillaSimoncelli metamer quality (#365), I ended up realizing I could make the code more efficient. This has been a goal for a while (see #222), and does involve making some improvements to steerable pyramid (#233 ) as well.

This PR largely consists of two major changes:
- I realized that the PortillaSimoncelli's autocorrelation computation could be converted to tensors (rather than lists of tensors) earlier, which makes things more efficient.
- The dictionary returned by `SteerablePyramidFreq.forward` no longer has `(scale, orientation)` keys with corresponding 4d (batch, channel, height, width) tensors, but `scale` keys with corresponding 5d (batch, channel, orientation, height, width) tensors. This is a **breaking change** but  faster on the GPU.

**Link any related issues, discussions, PRs**

This PR will close #222 for now -- I think further improvements in efficiency are possible, but they would require a more substantive reworking of the code (as described in that issue) and, as discussed in #365, the performance of the PortillaSimoncelli metamer synthesis is now a good deal better than the matlab code for a given synthesis duration (and continues to decrease beyond what the matlab can accomplish).

Also, as discussed below, I don't think pushing further on this will actually speed up metamer synthesis much. The duration of the steerable pyramid's forward pass is now sub-millisecond on the GPU, and PortillaSimoncelli's forward pass is now at about 10 milliseconds on the GPU. It's hard to push those much farther...

This does not close #233 (because I think the variables can be better named and `__init__`, at least, can be cleaned up), but does get it started.

**Outstanding questions / particular feedback**

I would mainly like input on how to communicate the changes to steerable pyramid effectively -- I have added  a `versionchanged` note to the docstring of the `forward` and `steer_coeffs` methods (as those are the only ones where the output has changed). Is that sufficient? should I also put a note in the main class docstring?

Also:
- For the dictionary-compatible non-linearities, those had been written largely with the steerable pyramid coefficients in mind. They now support both 5d (as the steerable pyramid returns) and 4d (as is standard), using vmap. ~The functions that do the same operation on individual tensors, rather than dictionaries (e.g., `local_gain_control`, as opposed to `local_gain_control_dict`) still only support 4d. Is that the right way to do this? Or should they also support 5d?~ It's largely about convenience, because the user could similarly use vmap to add support for extra dimensions -- should I just include an example of how to do that?
    - I've moved the 4d/5d support to the tensor level functions after Sarah's comments.
- Relatedly, currently which dimension we vmap over cannot be specified (it's fixed at `axis=2`, which is the extra "orientation" dimension for the steerable pyramid coefficients). Is that right? I'm not sure it makes a big difference -- I'm pretty sure that it only matters that height and width are on the last dimension, all others are handled independently.

**Describe changes**

- Changed how PortillaSimoncelli's autocorrelation computation happens.
- Changed how SteerablePyramidFreq's `forward` computation happens, changing the resulting shape of the tensors. **breaking change**.
- Changed other methods of SteerablePyramidFreq to be compatible with this change.
- Updated SteerablePyramidFreq's docstrings, typing, notebook, and tests to reflect this change. With the exception of testing against [pyrtools](https://github.com/LabForComputationalVision/pyrtools/) version (whose output we no longer match), this generally  made things cleaner.
- SteerablePyramidFreq's `steer_coeffs` method has been changed so that we now *only* output the new orientations, instead of concatenating the new ones to those in the input. I do not recall why we did this, but it seems like a weird decision. **breaking change** 
- SteerablePyramidFreq's `convert_pyr_to_tensor` and `convert_tensor_to_pyr` methods have been rewritten using `einops` instead of a bunch of for loops. I find the newer version easier to parse.
- Changed the non-linearities to support both 4d and 5d tensors (using vmap).
- Increased atol in the test for the Fisher vector product -- no change I made here should've affected this and indeed running the test on main on my workstation (in an environment that hasn't been updated!) also causes it to fail so ... not sure what changed. This only happens on GPU.
- Following #380, Jenkins CI started failing because of a pip ssl error. Following Dylan's suggestion that this was due to using an old cache, we now delete the Jenkins virtual environment at the end of every run to avoid this problem. This does mean that we're redownloading all our dependencies, but, given that we're using the pytorch image (which was the largest part of our virtual environment), this hopefully shouldn't add too much time.

**Checklist**

Affirm that you have done the following:

- [x] I have described the changes in this PR, following the template above.
- [x] I have added any necessary tests.
- [x] I have added any necessary documentation. This includes docstrings, updates to existing files found in `docs/`, or (for large changes) adding new files to the `docs/` folder.

**Benchmarking**

Some distribution plots to show the effect of this. In all of these, an individual plot is a specific function, main branch is in blue, this branch is in orange, device (cpu/gpu) are separated on the x-axis and y-axis shows the time in seconds (log-scaled):

time_steerablepyramid

For the steerable pyramid:
- `forward` on the CPU is the same, whereas GPU is ~5x faster
- same for `recon_pyr`
- `convert_pyr_to_tensor(split_complex=True)` is ~5x slower on cpu
- everything else is about ~5x faster.

time_portillasimoncelli

PortillaSimoncelli
- `forward` is about ~2x faster on gpu, same on cpu
- `compute_pyr_coeffs` is about 5x faster on gpu, 2x faster on cpu
- `reconstruct_lowpass` is about 2x faster on gpu, same on cpu

time_metamer

Metamer synthesis with 10 iterations:
- `Adam` and `LBFGS(history_size=3)` are abou 2x faster on GPU, slightly faster on CPU
- as history size increases, the speed up becomes smaller.
- from using py-spy (not shown here), I can see that as history size grows, more and more of the time is spent multiplying the (larger and larger) gradient history with itself and so there's diminishing returns from speeding up `PortillaSimoncelli.forward`. Given that my recommendation in #365 is to use `history_size=100`, I think pushing further on that is unlikely to be worthwhile.
- Note that, while 10 iterations of Adam synthesis is (much) faster than that of LBFGS, LBFGS requires far fewer iterations to find a good solution (i.e., the loss of synthesis with Adam is much higher than that of synthesis with LBFGS for the same number of iterations). With all these plots, one should only be looking within an axis, not across them. (For evidence of LBFGS doing better than Adam, see #365 )

(code and data for this can be found at [plenoptic-experiments/ps_speed_pr](https://github.com/plenoptic-org/plenoptic-experiments/tree/main/ps_speed_pr))

### CODE REVIEW

**Comment by billbrod** [comment]
Documentation built by flatiron-jenkins at http://docs.plenoptic.org/docs//pulls/377

File: docs/reproducibility.md [hunk 1]
```diff
@@ -44,11 +44,11 @@ The object you have loaded is a dictionary with strings as keys. Once you've loa
...
+A small change was made to the {class}`Metamer <plenoptic.synthesize.metamer.Metamer>` and {class}`MADCompetition <plenoptic.synthesize.mad_competition.MADCompetition>` APIs in `plenoptic` 1.3.1. You will be able to load {class}`Metamer <plenoptic.synthesize.metamer.Metamer>` and {class}`MADCompetition <plenoptic.synthesize.mad_competition.MADCompetition>` objects saved with version 1.2 and 1.3 for some time, but doing so will raise a `FutureWarning` and this compatibility will eventually be removed.
 
-In order to make an object compatible with future releases, you can either load it in with `plenoptic` 1.4 and re-save it, or do the following:
+In order to make an object compatible with future releases, you can either load it in with `plenoptic` 1.3.1 re-save it, or do the following:
```
Comment (sjvenditto) [inline docs/reproducibility.md]:
```suggestion
In order to make an object compatible with future releases, you can either load it in with `plenoptic` 1.3.1 and re-save it, or do the following:
```

**Review by BalzaniEdoardo — APPROVED** [review]
I think this LGTM. The only question I have is: do you have specific error messages for people that tries to use the steerable pyramid forward method with the tensor of the wrong shapes? i.e. users that updated plenoptic but are unaware of the breaking change?

I would mention specifically the the breaking change, so that it is immediately clear what is happening and how they should restructure their inputs

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -26,6 +26,10 @@ In a list, describe the changes you made. This should be more granular than the
 > - small change to synthesis to be compatible with new API
 > - while implementing this, a new version of torch was released, which required a small update to perceptual_distance.py
 
+**Benchmarking**
+
+If your test involves improvements to the efficiency of any part of the code base, please include some benchmarking results, to show the performance before (i.e., in the `main` branch) and after. See [PR #377](https://github.com/plenoptic-org/plenoptic/pull/377) for an example.
```
[the remark this answers is not in the document]
Comment (billbrod) [reply .github/pull_request_template.md]: Good call!

**Comment by billbrod** [comment]
> I think this LGTM. The only question I have is: do you have specific error messages for people that tries to use the steerable pyramid forward method with the tensor of the wrong shapes? i.e. users that updated plenoptic but are unaware of the breaking change?
> 
> I would mention specifically the the breaking change, so that it is immediately clear what is happening and how they should restructure their inputs

The output of the steerable pyramid forward method is changed, not the input. So I can't catch that for people. Right now, I have the `versionchanged` method in the docstring, and I'm going to add a big visible comment at the top of the release notes, but I don't think there's anything else I can do for folks...
